### PR TITLE
Initial check-in for the terms of purchase endpoint.

### PR DESCRIPTION
**Pull request by ryanopenenglish**

### CODE REVIEW

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java [hunk 1]
```diff
@@ -108,6 +117,46 @@ public ContractService() {
 
   }
 
+  public TermsOfPurchaseDTO createTermsOfPurchase(String country, String language, Integer initialTerm) {
+    TermsOfPurchaseDTO termsOfPurchaseDTO = new TermsOfPurchaseDTO();
+    DataModelAssembler dtm = new DataModelAssembler();
+    LocaleResolver localeResolver = LocaleResolver.EsEn;
+    if ("BR".equalsIgnoreCase(country)) {
```
Comment (johntran777) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: Might want to check with Chris to know what to use for "BR"

File: pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  private ContractService contractService;
+
+  @RequestMapping(method = RequestMethod.GET)
+  @ResponseBody //TODO not sure this is all that restful, throwing query params at this
+  public ResponseEntity getTermsOfPurchase(@RequestParam(value = "country", required = true) String country,
+      @RequestParam(value = "language",required = true) String language,
+      @RequestParam(value = "initialTerm", required = true) Integer initialTerm) {
+    try {
+      return new ResponseEntity(contractService.createTermsOfPurchase(country, language, initialTerm), HttpStatus.OK);
+    } catch (Exception e) { //TODO make this error handling a bit more robust / user friendly / error code appropriate
```
Comment (johntran777) [inline pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java]: Use the generic error handler, like AbstractVaultController or AbstractApiGatewayController.

File: pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  @Autowired
+  private ContractService contractService;
+
+  @RequestMapping(method = RequestMethod.GET)
+  @ResponseBody //TODO not sure this is all that restful, throwing query params at this
+  public ResponseEntity getTermsOfPurchase(@RequestParam(value = "country", required = true) String country,
+      @RequestParam(value = "language",required = true) String language,
+      @RequestParam(value = "initialTerm", required = true) Integer initialTerm) {
+    try {
+      return new ResponseEntity(contractService.createTermsOfPurchase(country, language, initialTerm), HttpStatus.OK);
```
Comment (johntran777) [inline pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java]: Return a 404 if there is not TOS for this.

Comment (johntran777) [inline pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java]: Will we have validation error -> BadRequest (400) error?

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java [hunk 4]
```diff
@@ -108,6 +117,46 @@ public ContractService() {
...
+      Template noRefundsTemplate = freemarkerConfiguration.getTemplate(noRefundsTemplateName);
+      moneyBackTOP = process(moneyBackGuaranteeTemplate, null);
+      noRefundsTOP = process(noRefundsTemplate, termsDataModel);
+    } catch (IOException e) {
+      LOGGER.error("Unable to retrieve Freemarker template: " + e.getMessage());
+      Throwables.propagate(e);
+    }
+
+    termsOfPurchaseDTO.getTermsOfPurchase().put(TOP_MONEY_BACK_GUARANTEE_TEMPLATE_NAME, moneyBackTOP);
+    termsOfPurchaseDTO.getTermsOfPurchase().put(TOP_NO_REFUNDS_TEMPLATE_NAME, noRefundsTOP);
```
Comment (johntran777) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: We only have these two TOP types?

File: pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/TermsOfPurchaseDTO.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.pojomatic.Pojomatic;
+import org.pojomatic.annotations.AutoProperty;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+@AutoProperty
+public class TermsOfPurchaseDTO {
```
Comment (johntran777) [inline pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/TermsOfPurchaseDTO.java]: Let's add @JsonIgnoreProperties(ignoreUnknown = true) and 
implements Serializable.

File: pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/DistributeContractDTO.java [hunk 6]
```diff
@@ -1,5 +1,9 @@
 package com.openenglish.pp.common.api.model;
 
+import org.pojomatic.Pojomatic;
+import org.pojomatic.annotations.AutoProperty;
+
+@AutoProperty
 public class DistributeContractDTO {
```
Comment (johntran777) [inline pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/DistributeContractDTO.java]: Let's add @JsonIgnoreProperties(ignoreUnknown = true) and 
implements Serializable.

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java [hunk 7]
```diff
@@ -108,6 +117,46 @@ public ContractService() {
 
   }
 
+  public TermsOfPurchaseDTO createTermsOfPurchase(String country, String language, Integer initialTerm) {
+    TermsOfPurchaseDTO termsOfPurchaseDTO = new TermsOfPurchaseDTO();
```
Comment (johntran777) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: Should we do Preconditions.checkArgument on the passed-in parameters (country, language, initialTerm)?

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java (hunk 4 above)
Comment (ryanopenenglish) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: There are others for financed orders, et.al., but this is all for upfront purchases.

File: pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java (hunk 3 above)
Comment (ryanopenenglish) [inline pp-product-catalog-webapp/src/main/java/com/openenglish/pp/controller/TermsOfPurchaseController.java]: Yeah, planning on doing validation / error handling tomorrow after I get the pipeline / demo sorted out.

File: pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/TermsOfPurchaseDTO.java (hunk 5 above)
Comment (ryanopenenglish) [inline pp-product-catalog-common/src/main/java/com/openenglish/pp/common/api/model/TermsOfPurchaseDTO.java]: Yeah, good call on the @JsonIgnoreProperties... why do you want Serializable? I can't think of any use for it...

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java (hunk 7 above)
Comment (ryanopenenglish) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: Yes, all validation will be added tomorrow.

File: pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java (hunk 1 above)
Comment (ryanopenenglish) [inline pp-product-catalog-service/src/main/java/com/openenglish/pp/productcatalog/service/ContractService.java]: Yeah, I forget how the enum debate shook out, so I put this in until I could ask someone.

**Comment by johntran777** [comment]
+1

**Pull request merged**